### PR TITLE
Sanity check for kpt/kustomize preinstalled version

### PR DESCRIPTION
**Pull request by yuwenma**
Fixes: #4884  #3904
**Merge after**: 4958

**Description**
Check kpt and kustomize CLI version to make sure it satisfies the basic requirements.

### CODE REVIEW

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 1]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
...
 }
 
+var sanityCheck = versionCheck
+
+// sanityCheck guarantees the kpt and kustomize versions are compatible with skaffold.
+func versionCheck(dir string) error {
+	kptCmd := exec.Command("kpt", "version")
+	out, err := util.RunCmdOut(kptCmd)
+	if err != nil {
+		return fmt.Errorf("running `kpt version` :%w", err)
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:87]: Can we make this error string more actionable. e.g.
fmt.Errorf("failed to determine kpt version due to %s. Try installing kpt again and try.", err)

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 2]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
...
+		// Kustomize version information is in the form of
+		// {Version:$VERSION GitCommit:$COMMIT BuildDate:1970-01-01T00:00:00Z GoOs:darwin GoArch:amd64}
+		re := regexp.MustCompile(kustomizeVersionRegexP)
+		match := re.FindStringSubmatch(versionInfo)
+		if len(match) != 2 {
+			return fmt.Errorf("unknown kustomize version %v\n Please upgrade your "+
+				"local kustomize CLI to a version >= %v\n  See kustomize installation: %v",
+				string(out), kustomizeMinVersion, kustomizeDownloadLink)
+		}
+		fmt.Printf("!!!!!!! kustomize %v min %v\n\n", match[1], kustomizeMinVersion)
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:120]: this looks unintentional

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 3]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
...
+		if !semver.IsValid(match[1]) || semver.Compare(match[1], kustomizeMinVersion) < 0 {
+			return fmt.Errorf("you are using kustomize %q\n"+
+				"Please update your kustomize version to  >= %v\n"+
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:123]: Looks like there is an  extra space "version to   >="

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 4]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
...
+				"  See kustomize installation: %v", match[1], kustomizeMinVersion,
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:124]: Since this on new line, we can remove the beginning " "

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 5]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
...
+		if err != nil {
+			return fmt.Errorf("running `kustomize version` :%w", err)
+		}
+		versionInfo := strings.TrimSuffix(string(out), "\n")
+		// Kustomize version information is in the form of
+		// {Version:$VERSION GitCommit:$COMMIT BuildDate:1970-01-01T00:00:00Z GoOs:darwin GoArch:amd64}
+		re := regexp.MustCompile(kustomizeVersionRegexP)
+		match := re.FindStringSubmatch(versionInfo)
+		if len(match) != 2 {
+			return fmt.Errorf("unknown kustomize version %v\n Please upgrade your "+
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:116]: Remove  space after new line char

File: pkg/skaffold/deploy/kpt/kpt.go [hunk 6]
```diff
@@ -69,10 +77,64 @@ func NewDeployer(cfg types.Config, labels map[string]string) *Deployer {
 	}
 }
 
+var sanityCheck = versionCheck
+
+// sanityCheck guarantees the kpt and kustomize versions are compatible with skaffold.
+func versionCheck(dir string) error {
```
Comment (tejal29) [inline pkg/skaffold/deploy/kpt/kpt.go:83]: Would it be possible for you to list the error messages in the test case or verify if they look ok manually?

**Review by tejal29 — CHANGES_REQUESTED** [review]
LGTM expect for minor nits on error strings.

File: pkg/skaffold/deploy/kpt/kpt.go (hunk 5 above)
Comment (yuwenma) [reply pkg/skaffold/deploy/kpt/kpt.go:116]: Done

File: pkg/skaffold/deploy/kpt/kpt.go (hunk 3 above)
Comment (yuwenma) [reply pkg/skaffold/deploy/kpt/kpt.go:123]: good catch

File: pkg/skaffold/deploy/kpt/kpt.go (hunk 6 above)
Comment (yuwenma) [reply pkg/skaffold/deploy/kpt/kpt.go:83]: Done

File: pkg/skaffold/deploy/kpt/kpt.go (hunk 1 above)
Comment (yuwenma) [reply pkg/skaffold/deploy/kpt/kpt.go:87]: Done

**Review by tejal29 — APPROVED** [review]

**Comment by MarlonGamez** [comment]
Looks good, I'll merge the changes once kokoro is done

**Comment by tejal29** [comment]
Manually running kokoro here https://fusion.corp.google.com/invocation/18f9d0a4-3cd2-4137-bbfd-779da58d200b?target=skaffold%2Fpresubmit&drilldownTab=tests

**Pull request merged**